File: mobile_client/android/NetworkClient.kt
package com.metin2.pvp.network

import android.util.Log
import okhttp3.*
import okio.ByteString
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.TimeUnit

/**
 * Metin2 PvP Network Client for Android
 * Hibrit protokol: WebSocket + UDP
 */
class NetworkClient(
    private val serverIP: String,
    private val wsPort: Int = 8080,
    private val udpPort: Int = 8081
) {
    companion object {
        private const val TAG = "NetworkClient"

        // Packet types
        const val HEADER_CG_LOGIN: Byte = 2
        const val HEADER_GC_LOGIN_SUCCESS: Byte = 3
        const val HEADER_CG_CHARACTER_SELECT: Byte = 10
        const val HEADER_GC_CHARACTER_INFO: Byte = 11
        const val HEADER_CG_MOVE: Byte = 20
        const val HEADER_GC_MOVE: Byte = 21
        const val HEADER_CG_ATTACK: Byte = 30
        const val HEADER_GC_DAMAGE: Byte = 32
        const val HEADER_CG_CHAT: Byte = 40
        const val HEADER_GC_CHAT: Byte = 41
        const val HEADER_CG_PING: Byte = 50
        const val HEADER_GC_PONG: Byte = 51
    }

    // WebSocket
    private var webSocket: WebSocket? = null
    private val okHttpClient = OkHttpClient.Builder()
        .pingInterval(15, TimeUnit.SECONDS)
        .build()

    // UDP
    private var udpSocket: DatagramSocket? = null
    private var udpThread: Thread? = null
    private var isUDPRunning = false
    private var udpAddress: InetAddress? = null

    // State
    private var isConnected = false
    private var sessionId: Int = 0
    private var characterId: Int = 0
    private var sequenceNumber: Int = 0

    // Callbacks
    var onConnected: (() -> Unit)? = null
    var onDisconnected: ((String) -> Unit)? = null
    var onPacketReceived: ((Byte, ByteArray) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    /**
     * WebSocket'e bağlan
     */
    fun connectWebSocket() {
        val url = "ws://$serverIP:$wsPort"
        val request = Request.Builder()
            .url(url)
            .build()

        webSocket = okHttpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.i(TAG, "WebSocket connected")
                isConnected = true
                onConnected?.invoke()
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                handleWebSocketMessage(bytes.toByteArray())
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.i(TAG, "WebSocket closing: $reason")
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.i(TAG, "WebSocket closed: $reason")
                isConnected = false
                onDisconnected?.invoke(reason)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket error: ${t.message}")
                isConnected = false
                onError?.invoke(t.message ?: "Unknown error")
            }
        })
    }

    /**
     * UDP başlat
     */
    fun startUDP() {
        try {
            udpSocket = DatagramSocket()
            udpAddress = InetAddress.getByName(serverIP)
            isUDPRunning = true

            // UDP receive thread
            udpThread = Thread {
                val buffer = ByteArray(2048)
                while (isUDPRunning) {
                    try {
                        val packet = DatagramPacket(buffer, buffer.size)
                        udpSocket?.receive(packet)

                        val data = packet.data.copyOfRange(0, packet.length)
                        handleUDPMessage(data)
                    } catch (e: Exception) {
                        if (isUDPRunning) {
                            Log.e(TAG, "UDP receive error: ${e.message}")
                        }
                    }
                }
            }.apply { start() }

            Log.i(TAG, "UDP started on port $udpPort")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start UDP: ${e.message}")
            onError?.invoke("UDP init failed")
        }
    }

    /**
     * Bağlantıyı kes
     */
    fun disconnect() {
        // WebSocket kapat
        webSocket?.close(1000, "Client disconnect")
        webSocket = null

        // UDP kapat
        isUDPRunning = false
        udpSocket?.close()
        udpThread?.join(1000)

        isConnected = false
        Log.i(TAG, "Disconnected")
    }

    // ========== PACKET SENDING ==========

    /**
     * Login paketi gönder
     */
    fun sendLogin(username: String, password: String) {
        val buffer = ByteBuffer.allocate(256).order(ByteOrder.LITTLE_ENDIAN)

        // Header
        buffer.put(HEADER_CG_LOGIN)
        buffer.putShort((8 + 56).toShort()) // size
        buffer.putInt(++sequenceNumber)
        buffer.putInt(System.currentTimeMillis().toInt())

        // Body
        buffer.put(username.toByteArray().copyOf(24))
        buffer.put(password.toByteArray().copyOf(32))

        sendWebSocket(buffer.array(), buffer.position())
    }

    /**
     * Karakter seçim paketi
     */
    fun sendCharacterSelect(charId: Int) {
        val buffer = ByteBuffer.allocate(64).order(ByteOrder.LITTLE_ENDIAN)

        buffer.put(HEADER_CG_CHARACTER_SELECT)
        buffer.putShort((8 + 4).toShort())
        buffer.putInt(++sequenceNumber)
        buffer.putInt(System.currentTimeMillis().toInt())
        buffer.putInt(charId)

        characterId = charId
        sendWebSocket(buffer.array(), buffer.position())
    }

    /**
     * Hareket paketi (UDP)
     */
    fun sendMove(x: Int, y: Int, dir: Byte) {
        val buffer = ByteBuffer.allocate(64).order(ByteOrder.LITTLE_ENDIAN)

        buffer.put(HEADER_CG_MOVE)
        buffer.putShort((8 + 9).toShort())
        buffer.putInt(++sequenceNumber)
        buffer.putInt(System.currentTimeMillis().toInt())

        buffer.putInt(x)
        buffer.putInt(y)
        buffer.put(dir)

        sendUDP(buffer.array(), buffer.position())
    }

    /**
     * Saldırı paketi (UDP)
     */
    fun sendAttack(targetId: Int) {
        val buffer = ByteBuffer.allocate(64).order(ByteOrder.LITTLE_ENDIAN)

        buffer.put(HEADER_CG_ATTACK)
        buffer.putShort((8 + 6).toShort())
        buffer.putInt(++sequenceNumber)
        buffer.putInt(System.currentTimeMillis().toInt())

        buffer.putInt(targetId)
        buffer.put(0) // attack_type
        buffer.putShort(0) // skill_id

        sendUDP(buffer.array(), buffer.position())
    }

    /**
     * Chat paketi
     */
    fun sendChat(message: String) {
        val buffer = ByteBuffer.allocate(512).order(ByteOrder.LITTLE_ENDIAN)

        buffer.put(HEADER_CG_CHAT)
        buffer.putShort((8 + 261).toShort())
        buffer.putInt(++sequenceNumber)
        buffer.putInt(System.currentTimeMillis().toInt())

        buffer.put(0) // type
        buffer.putInt(0) // target_id
        buffer.put(message.toByteArray().copyOf(256))

        sendWebSocket(buffer.array(), buffer.position())
    }

    /**
     * Ping paketi
     */
    fun sendPing() {
        val buffer = ByteBuffer.allocate(32).order(ByteOrder.LITTLE_ENDIAN)

        val clientTime = System.currentTimeMillis().toInt()

        buffer.put(HEADER_CG_PING)
        buffer.putShort((8 + 4).toShort())
        buffer.putInt(++sequenceNumber)
        buffer.putInt(clientTime)
        buffer.putInt(clientTime)

        sendWebSocket(buffer.array(), buffer.position())
    }

    // ========== LOW-LEVEL SEND ==========

    private fun sendWebSocket(data: ByteArray, size: Int) {
        webSocket?.send(ByteString.of(data, 0, size))
    }

    private fun sendUDP(data: ByteArray, size: Int) {
        try {
            val packet = DatagramPacket(data, size, udpAddress, udpPort)
            udpSocket?.send(packet)
        } catch (e: Exception) {
            Log.e(TAG, "UDP send error: ${e.message}")
        }
    }

    // ========== PACKET HANDLING ==========

    private fun handleWebSocketMessage(data: ByteArray) {
        if (data.size < 8) return

        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val type = buffer.get()
        val size = buffer.short.toInt()
        val sequence = buffer.int
        val timestamp = buffer.int

        Log.d(TAG, "WS Packet: type=$type, size=$size, seq=$sequence")

        val body = ByteArray(size - 8)
        buffer.get(body)

        onPacketReceived?.invoke(type, body)

        // Specific handlers
        when (type) {
            HEADER_GC_LOGIN_SUCCESS -> handleLoginSuccess(body)
            HEADER_GC_CHARACTER_INFO -> handleCharacterInfo(body)
            HEADER_GC_CHAT -> handleChat(body)
            HEADER_GC_PONG -> handlePong(body)
        }
    }

    private fun handleUDPMessage(data: ByteArray) {
        if (data.size < 8) return

        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val type = buffer.get()
        val size = buffer.short.toInt()
        val sequence = buffer.int
        val timestamp = buffer.int

        Log.d(TAG, "UDP Packet: type=$type, size=$size")

        val body = ByteArray(size - 8)
        buffer.get(body)

        onPacketReceived?.invoke(type, body)

        // Specific handlers
        when (type) {
            HEADER_GC_MOVE -> handleMove(body)
            HEADER_GC_DAMAGE -> handleDamage(body)
        }
    }

    private fun handleLoginSuccess(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val accountId = buffer.int
        val charCount = buffer.get()

        Log.i(TAG, "Login success! Account: $accountId, Characters: $charCount")
    }

    private fun handleCharacterInfo(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val id = buffer.int
        val name = String(ByteArray(24).apply { buffer.get(this) }).trim('\u0000')
        val job = buffer.get()
        val level = buffer.get()
        val hp = buffer.int
        val maxHp = buffer.int

        Log.i(TAG, "Character: $name (ID:$id) Lv.$level HP:$hp/$maxHp")
    }

    private fun handleChat(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val senderId = buffer.int
        val senderName = String(ByteArray(24).apply { buffer.get(this) }).trim('\u0000')
        val type = buffer.get()
        val message = String(ByteArray(256).apply { buffer.get(this) }).trim('\u0000')

        Log.i(TAG, "[CHAT] $senderName: $message")
    }

    private fun handlePong(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val clientTime = buffer.int
        val serverTime = buffer.int
        val latency = System.currentTimeMillis().toInt() - clientTime

        Log.d(TAG, "Pong received. Latency: ${latency}ms")
    }

    private fun handleMove(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val id = buffer.int
        val x = buffer.int
        val y = buffer.int
        val dir = buffer.get()

        Log.d(TAG, "Player $id moved to ($x, $y)")
    }

    private fun handleDamage(data: ByteArray) {
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        val attackerId = buffer.int
        val victimId = buffer.int
        val damage = buffer.int
        val isCritical = buffer.get()

        Log.i(TAG, "Damage: $attackerId -> $victimId = $damage${if (isCritical.toInt() == 1) " CRIT!" else ""}")
    }
}

/**
 * Kullanım örneği
 */
class NetworkExample {
    fun example() {
        val client = NetworkClient("192.168.1.100")

        client.onConnected = {
            println("Bağlandı!")
            client.sendLogin("testuser", "password123")
        }

        client.onDisconnected = { reason ->
            println("Bağlantı kesildi: $reason")
        }

        client.onPacketReceived = { type, data ->
            println("Paket alındı: type=$type, size=${data.size}")
        }

        client.connectWebSocket()
        client.startUDP()

        // Hareket gönderme örneği
        Thread {
            Thread.sleep(2000)
            client.sendMove(957300, 245000, 0)
        }.start()

        // Ping gönderme örneği
        Thread {
            while (true) {
                Thread.sleep(5000)
                client.sendPing()
            }
        }.start()
    }
}

File: mobile_client/ios/NetworkClient.swift
import Foundation
import Network

/**
 * Metin2 PvP Network Client for iOS
 * Hibrit protokol: WebSocket + UDP
 * Swift 5.0+, iOS 13.0+
 */

class NetworkClient {

    // MARK: - Constants

    enum PacketType: UInt8 {
        case cgLogin = 2
        case gcLoginSuccess = 3
        case cgCharacterSelect = 10
        case gcCharacterInfo = 11
        case cgMove = 20
        case gcMove = 21
        case cgAttack = 30
        case gcDamage = 32
        case cgChat = 40
        case gcChat = 41
        case cgPing = 50
        case gcPong = 51
    }

    // MARK: - Properties

    private let serverIP: String
    private let wsPort: Int
    private let udpPort: Int

    // WebSocket
    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?

    // UDP (Using Network.framework)
    private var udpConnection: NWConnection?
    private let udpQueue = DispatchQueue(label: "com.metin2.udp")

    // State
    private var isConnected = false
    private var sessionId: UInt32 = 0
    private var characterId: UInt32 = 0
    private var sequenceNumber: UInt32 = 0

    // Callbacks
    var onConnected: (() -> Void)?
    var onDisconnected: ((String) -> Void)?
    var onPacketReceived: ((UInt8, Data) -> Void)?
    var onError: ((String) -> Void)?

    // MARK: - Initialization

    init(serverIP: String, wsPort: Int = 8080, udpPort: Int = 8081) {
        self.serverIP = serverIP
        self.wsPort = wsPort
        self.udpPort = udpPort
    }

    // MARK: - Connection

    func connectWebSocket() {
        let urlString = "ws://\(serverIP):\(wsPort)"
        guard let url = URL(string: urlString) else {
            onError?("Invalid URL")
            return
        }

        urlSession = URLSession(configuration: .default)
        webSocketTask = urlSession?.webSocketTask(with: url)

        webSocketTask?.resume()
        isConnected = true
        onConnected?()

        // Start receiving messages
        receiveWebSocketMessage()

        print("[WS] Connected to \(urlString)")
    }

    func startUDP() {
        let host = NWEndpoint.Host(serverIP)
        let port = NWEndpoint.Port(integerLiteral: UInt16(udpPort))

        udpConnection = NWConnection(host: host, port: port, using: .udp)

        udpConnection?.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                print("[UDP] Connected")
                self?.receiveUDPMessage()

            case .failed(let error):
                print("[UDP] Failed: \(error)")
                self?.onError?("UDP connection failed")

            case .cancelled:
                print("[UDP] Cancelled")

            default:
                break
            }
        }

        udpConnection?.start(queue: udpQueue)
    }

    func disconnect() {
        // Close WebSocket
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        urlSession = nil

        // Close UDP
        udpConnection?.cancel()
        udpConnection = nil

        isConnected = false
        print("[Network] Disconnected")
    }

    // MARK: - Packet Sending

    func sendLogin(username: String, password: String) {
        var buffer = Data()

        // Header
        buffer.append(PacketType.cgLogin.rawValue)
        buffer.append(UInt16(8 + 56).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(UInt32(Date().timeIntervalSince1970 * 1000).littleEndianData)

        // Body
        buffer.append(username.paddedData(to: 24))
        buffer.append(password.paddedData(to: 32))

        sendWebSocket(data: buffer)
    }

    func sendCharacterSelect(charId: UInt32) {
        var buffer = Data()

        buffer.append(PacketType.cgCharacterSelect.rawValue)
        buffer.append(UInt16(8 + 4).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(UInt32(Date().timeIntervalSince1970 * 1000).littleEndianData)
        buffer.append(charId.littleEndianData)

        characterId = charId
        sendWebSocket(data: buffer)
    }

    func sendMove(x: Int32, y: Int32, dir: UInt8) {
        var buffer = Data()

        buffer.append(PacketType.cgMove.rawValue)
        buffer.append(UInt16(8 + 9).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(UInt32(Date().timeIntervalSince1970 * 1000).littleEndianData)

        buffer.append(x.littleEndianData)
        buffer.append(y.littleEndianData)
        buffer.append(dir)

        sendUDP(data: buffer)
    }

    func sendAttack(targetId: UInt32) {
        var buffer = Data()

        buffer.append(PacketType.cgAttack.rawValue)
        buffer.append(UInt16(8 + 6).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(UInt32(Date().timeIntervalSince1970 * 1000).littleEndianData)

        buffer.append(targetId.littleEndianData)
        buffer.append(UInt8(0)) // attack_type
        buffer.append(UInt16(0).littleEndianData) // skill_id

        sendUDP(data: buffer)
    }

    func sendChat(message: String) {
        var buffer = Data()

        buffer.append(PacketType.cgChat.rawValue)
        buffer.append(UInt16(8 + 261).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(UInt32(Date().timeIntervalSince1970 * 1000).littleEndianData)

        buffer.append(UInt8(0)) // type
        buffer.append(UInt32(0).littleEndianData) // target_id
        buffer.append(message.paddedData(to: 256))

        sendWebSocket(data: buffer)
    }

    func sendPing() {
        var buffer = Data()

        let clientTime = UInt32(Date().timeIntervalSince1970 * 1000)

        buffer.append(PacketType.cgPing.rawValue)
        buffer.append(UInt16(8 + 4).littleEndianData)
        sequenceNumber += 1
        buffer.append(sequenceNumber.littleEndianData)
        buffer.append(clientTime.littleEndianData)
        buffer.append(clientTime.littleEndianData)

        sendWebSocket(data: buffer)
    }

    // MARK: - Low-level Send

    private func sendWebSocket(data: Data) {
        let message = URLSessionWebSocketTask.Message.data(data)
        webSocketTask?.send(message) { error in
            if let error = error {
                print("[WS] Send error: \(error)")
                self.onError?("WebSocket send failed")
            }
        }
    }

    private func sendUDP(data: Data) {
        udpConnection?.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                print("[UDP] Send error: \(error)")
            }
        })
    }

    // MARK: - Receiving

    private func receiveWebSocketMessage() {
        webSocketTask?.receive { [weak self] result in
            guard let self = self else { return }

            switch result {
            case .success(let message):
                switch message {
                case .data(let data):
                    self.handleWebSocketMessage(data: data)

                case .string(let text):
                    print("[WS] Received text: \(text)")

                @unknown default:
                    break
                }

                // Continue receiving
                self.receiveWebSocketMessage()

            case .failure(let error):
                print("[WS] Receive error: \(error)")
                self.onDisconnected?("WebSocket error")
            }
        }
    }

    private func receiveUDPMessage() {
        udpConnection?.receiveMessage { [weak self] data, context, isComplete, error in
            guard let self = self else { return }

            if let data = data, !data.isEmpty {
                self.handleUDPMessage(data: data)
            }

            if let error = error {
                print("[UDP] Receive error: \(error)")
            } else {
                // Continue receiving
                self.receiveUDPMessage()
            }
        }
    }

    // MARK: - Packet Handling

    private func handleWebSocketMessage(data: Data) {
        guard data.count >= 8 else { return }

        let type = data[0]
        let size = data.readUInt16(at: 1)
        let sequence = data.readUInt32(at: 3)
        let timestamp = data.readUInt32(at: 7)

        print("[WS] Packet: type=\(type), size=\(size), seq=\(sequence)")

        let body = data.subdata(in: 8..<data.count)
        onPacketReceived?(type, body)

        // Specific handlers
        switch type {
        case PacketType.gcLoginSuccess.rawValue:
            handleLoginSuccess(data: body)
        case PacketType.gcCharacterInfo.rawValue:
            handleCharacterInfo(data: body)
        case PacketType.gcChat.rawValue:
            handleChat(data: body)
        case PacketType.gcPong.rawValue:
            handlePong(data: body)
        default:
            break
        }
    }

    private func handleUDPMessage(data: Data) {
        guard data.count >= 8 else { return }

        let type = data[0]
        let size = data.readUInt16(at: 1)

        print("[UDP] Packet: type=\(type), size=\(size)")

        let body = data.subdata(in: 8..<data.count)
        onPacketReceived?(type, body)

        // Specific handlers
        switch type {
        case PacketType.gcMove.rawValue:
            handleMove(data: body)
        case PacketType.gcDamage.rawValue:
            handleDamage(data: body)
        default:
            break
        }
    }

    private func handleLoginSuccess(data: Data) {
        let accountId = data.readUInt32(at: 0)
        let charCount = data[4]

        print("[Login] Success! Account: \(accountId), Characters: \(charCount)")
    }

    private func handleCharacterInfo(data: Data) {
        let id = data.readUInt32(at: 0)
        let name = data.readString(at: 4, length: 24)
        let job = data[28]
        let level = data[29]
        let hp = data.readUInt32(at: 30)
        let maxHp = data.readUInt32(at: 34)

        print("[Character] \(name) (ID:\(id)) Lv.\(level) HP:\(hp)/\(maxHp)")
    }

    private func handleChat(data: Data) {
        let senderId = data.readUInt32(at: 0)
        let senderName = data.readString(at: 4, length: 24)
        let type = data[28]
        let message = data.readString(at: 29, length: 256)

        print("[CHAT] \(senderName): \(message)")
    }

    private func handlePong(data: Data) {
        let clientTime = data.readUInt32(at: 0)
        let serverTime = data.readUInt32(at: 4)
        let latency = UInt32(Date().timeIntervalSince1970 * 1000) - clientTime

        print("[Pong] Latency: \(latency)ms")
    }

    private func handleMove(data: Data) {
        let id = data.readUInt32(at: 0)
        let x = data.readInt32(at: 4)
        let y = data.readInt32(at: 8)
        let dir = data[12]

        print("[Move] Player \(id) moved to (\(x), \(y))")
    }

    private func handleDamage(data: Data) {
        let attackerId = data.readUInt32(at: 0)
        let victimId = data.readUInt32(at: 4)
        let damage = data.readUInt32(at: 8)
        let isCritical = data[12]

        let critText = isCritical == 1 ? " CRIT!" : ""
        print("[Damage] \(attackerId) -> \(victimId) = \(damage)\(critText)")
    }
}

// MARK: - Data Extensions

extension Data {
    func readUInt8(at offset: Int) -> UInt8 {
        return self[offset]
    }

    func readUInt16(at offset: Int) -> UInt16 {
        return withUnsafeBytes { $0.load(fromByteOffset: offset, as: UInt16.self) }
    }

    func readUInt32(at offset: Int) -> UInt32 {
        return withUnsafeBytes { $0.load(fromByteOffset: offset, as: UInt32.self) }
    }

    func readInt32(at offset: Int) -> Int32 {
        return withUnsafeBytes { $0.load(fromByteOffset: offset, as: Int32.self) }
    }

    func readString(at offset: Int, length: Int) -> String {
        let subdata = self.subdata(in: offset..<(offset + length))
        // Remove null terminators
        let cleanData = subdata.prefix(while: { $0 != 0 })
        return String(data: cleanData, encoding: .utf8) ?? ""
    }
}

extension String {
    func paddedData(to length: Int) -> Data {
        var data = self.data(using: .utf8) ?? Data()
        if data.count < length {
            data.append(Data(repeating: 0, count: length - data.count))
        } else if data.count > length {
            data = data.prefix(length)
        }
        return data
    }
}

extension FixedWidthInteger {
    var littleEndianData: Data {
        var value = self.littleEndian
        return Data(bytes: &value, count: MemoryLayout<Self>.size)
    }
}

// MARK: - Usage Example

class NetworkExample {
    let client = NetworkClient(serverIP: "192.168.1.100")

    func example() {
        client.onConnected = {
            print("Connected!")
            self.client.sendLogin(username: "testuser", password: "password123")
        }

        client.onDisconnected = { reason in
            print("Disconnected: \(reason)")
        }

        client.onPacketReceived = { type, data in
            print("Packet received: type=\(type), size=\(data.count)")
        }

        client.connectWebSocket()
        client.startUDP()

        // Movement example
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            self.client.sendMove(x: 957300, y: 245000, dir: 0)
        }

        // Ping example
        Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { _ in
            self.client.sendPing()
        }
    }
}
